pkg/policycache/cache.go

```diff
@@ -89,131 +91,117 @@ func (m *pMap) add(policy *kyverno.ClusterPolicy) {
 		pName = pSpace + "/" + pName
 		isNamespacedPolicy = true
 		// Initialize Namespace Cache Map
-		_, ok := m.nsDataMap[policy.GetNamespace()]
-		if !ok {
-			m.nsDataMap[policy.GetNamespace()] = make(map[PolicyType][]*kyverno.ClusterPolicy)
-		}
 	}
-
 	for _, rule := range policy.Spec.Rules {
-		if rule.HasMutate() {
-			if !mutateMap[pName] {
-				mutateMap[pName] = true
-				if isNamespacedPolicy {
-					mutatePolicy := m.nsDataMap[policy.GetNamespace()][Mutate]
-					m.nsDataMap[policy.GetNamespace()][Mutate] = append(mutatePolicy, policy)
-					continue
-				}
-				mutatePolicy := m.dataMap[Mutate]
-				m.dataMap[Mutate] = append(mutatePolicy, policy)
+
+		for _, kind := range rule.MatchResources.Kinds {
+			_, ok := m.kindDataMap[kind]
+			if !ok {
+				m.kindDataMap[kind] = make(map[PolicyType][]string)
 			}
-			continue
-		}
 
-		if rule.HasValidate() {
-			if enforcePolicy {
-				if !validateEnforceMap[pName] {
-					validateEnforceMap[pName] = true
+			if rule.HasMutate() {
+				if !mutateMap[kind+"/"+pName] {
+					mutateMap[kind+"/"+pName] = true
 					if isNamespacedPolicy {
-						validatePolicy := m.nsDataMap[policy.GetNamespace()][ValidateEnforce]
-						m.nsDataMap[policy.GetNamespace()][ValidateEnforce] = append(validatePolicy, policy)
+						mutatePolicy := m.kindDataMap[kind][Mutate]
+						m.kindDataMap[kind][Mutate] = append(mutatePolicy, pName)
 						continue
 					}
-					validatePolicy := m.dataMap[ValidateEnforce]
-					m.dataMap[ValidateEnforce] = append(validatePolicy, policy)
+					mutatePolicy := m.kindDataMap[kind][Mutate]
+					m.kindDataMap[kind][Mutate] = append(mutatePolicy, policy.GetName())
```

Comment:
These 2 blocks look duplicate, if we already use kind/namespace/name as the map key, there's no need for the `if isNamespacedPolicy {}` block, pName is already concatenated for nsPol:
```
pName = pSpace + "/" + pName
```